br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -573,6 +587,100 @@ func (m *DupeDetector) RecordIndexConflictError(ctx context.Context, stream DupK
 	return nil
 }
 
+// RetrieveKeyAndValueFromErrFoundDuplicateKeys retrieves the key and value
+// from ErrFoundDuplicateKeys error.
+func RetrieveKeyAndValueFromErrFoundDuplicateKeys(err error) ([]byte, []byte, error) {
+	if !common.ErrFoundDuplicateKeys.Equal(err) {
+		return nil, nil, err
+	}
+	tErr, ok := errors.Cause(err).(*terror.Error)
+	if !ok {
+		return nil, nil, err
+	}
+	if len(tErr.Args()) != 2 {
+		return nil, nil, err
+	}
+	key, keyIsByte := tErr.Args()[0].([]byte)
+	value, valIsByte := tErr.Args()[1].([]byte)
+	if !keyIsByte || !valIsByte {
+		return nil, nil, err
+	}
+	return key, value, nil
+}
+
+// NewErrFoundConflictRecords generate an error ErrFoundDataConflictRecords / ErrFoundIndexConflictRecords
+// according to key and value.
+func NewErrFoundConflictRecords(key []byte, value []byte, tbl table.Table) error {
```

Comment:
why not create the error in where `ErrFoundDuplicateKeys` is created, so no need to convert them back and forth

Reply:
Because `ErrFoundDuplicateKeys` is created here
https://github.com/pingcap/tidb/blob/69ab2a4de69f458db967026c7da5da0bbfba1c79/br/pkg/lightning/common/dupdetect.go#L96-L102
The error will be returned to lightning/ddl, and we need to maintain the behavior of ddl, which will convert `ErrFoundDuplicateKeys` to `ErrKeyExists`. But in lightning, we will need to convert `ErrFoundDuplicateKeys` to `ErrFoundDataConflictRecords` or `ErrFoundIndexConflictRecords` for better readability according to PM, so we do separate conversion in lightning.
So actually there are no converting back and forth, there are just two paths for converting depending on which package the error is returned to.

Reply:
can we return `ErrFoundDataConflictRecords` and `ErrFoundIndexConflictRecords ` directly, and change ddl part to convert `ErrFoundIndexConflictRecords` to `ErrKeyExists`

Reply:
The original `ErrFoundDuplicateKeys` error contains the most basic information of the conflict. When creating `ErrFoundDataConflictRecords` or `ErrFoundIndexConflictRecords`, we need to conduct extra decoding and encoding for information like row key, row value, etc. 
`ErrKeyExists` does not need those extra information, so it seems unnecessary to convert `ErrKeyExists` from `ErrFoundIndexConflictRecords`.